# Programming in Python

## Conditioning (If)

In [7]:
NamesPython = ["Michelangelo", "Roxane", "Alienor", "Lou", "Arrigo", "Anabelle", "Irene", "Celia", "Carmen", "Oriane", "Sofia"]

In [8]:
minombre = input("¿Cómo te llamas?")

¿Cómo te llamas?Michelangelo


In [9]:
if minombre in NamesPython:
  print("¡Bienvenido al curso!")
else:
  print("Je suis desolé!, no estás en la lista")

¡Bienvenido al curso!


**Example: Grading**

Divide in 3 parts:



In [10]:
numerical = 75

if numerical > 90:
  grade = "A"
elif numerical >= 75 and numerical <= 90:
  grade = "B"
elif numerical >= 50 and numerical < 75:
  grade = "C"
else:
  grade = "F"

print("Your grade is " + grade)


Your grade is B


## Loops

"Write the first 100 numbers."

There are several assumptions implicit in this seemingly simple command. The first is
that the student knows where to start, namely at number 1. The second assumption is
that the student knows where to end, at number 100. And the third is that the student
understands that they should count by 1.

In programming, this set of instructions may be executed with a loop.

There are three key components to most loops:
1. The start of the loop
2. The end of the loop
3. The increment between numbers in the loop

Python distinguishes between two fundamental kinds of loops: **while** loops, and **for**
loops.

### While

In a while loop, a designated segment of code repeats provided that a particular
condition is true. When the condition evaluates to false, the while loop stops running.
The while loops print out the first 10 numbers.

You could print the first 10 numbers by implementing the print function 10 times, but
using a while loop is more efficient, and it scales easily. In general, it's not a good idea to
copy and paste while coding. If you find yourself copying and pasting, there's probably a
more efficient way. Let's have a look at the following example code block:

In [11]:
i = 1
while i <= 10:
  print(i)
  i += 1


1
2
3
4
5
6
7
8
9
10


You can break down the preceding code block and find out what's happening in
concrete steps:



*   Initialize the variable: Loops need to be initialized with a variable. The variable is
going to change throughout the loop. The naming of the variable is up to you. i is
often chosen because it stands for incrementor. An example is i = 1.
*   Set up the while loop: The while loop starts with the while keyword. Following
while is the chosen variable. After the variable comes the condition that must be
met for the loop to run. In general, the condition should have some way of being
broken. When counting, the condition usually includes an upper limit, but it
could also be broken in other ways, such as i != 10. This line of code is the most
critical piece of the loop. It sets up how many times the loop is expected to run. An
example is while i <= 10:.
*   Instructions: The instructions include all indented lines after the colon. Anything
could be printed, any function could be called, and any number of lines may
be executed. It all depends on the program. As long as the code is syntactically
correct, generally speaking, anything goes. This part of the loop is going to run
over and over as long as the aforementioned condition is true. An example is
print(i).
*   Increment: The incrementor is a crucial part of this example. Without it, the
preceding code will never stop running. It will print 1's endlessly because 1 is
always less than 10. Here, you increment by 1, but you could also increment by 2,
or any other number. An example is i += 1.



**break** is a special keyword in Python that is specifically designed for loops. If placed
inside of a loop, commonly in a conditional, break will immediately terminate the loop. It
doesn't matter what comes before or after the loop. The break is placed on its own line,
and it breaks out of the loop.

To practice, you should print the first number greater than 100 that is divisible by 17.

The idea is that you are going to start at 101 and keep counting until you find a number
divisible by 17. Assume you don't know what number to stop at. This is where break
comes into play. break will terminate the loop. You can set our upper bound at some
number that you know you will never reach and break out of the loop when you get
there:

In [12]:
x = 100
while x <= 1000:
  x += 1
  if x%17==0:
    print('', x, 'is the first number greater than 100 that is divisible by 17.')
    break

 102 is the first number greater than 100 that is divisible by 17.


The `x += 1` iterator is placed at the beginning of the loop. This allows us to start with 101. The iterator may be placed anywhere in the loop.

Since 101 is not divisible by 17, the loop repeats, and x = 102. Since 102 is divisible by 17,
the print statement executes and we break out of the loop.

This is the first time you have used double indentation. Since the if conditional is
inside of a while loop, it must be indented as well.

#### Exercise LCM

In this activity, you will find the LCM of two divisors. The LCM of two divisors is the first
number that both divisors can divide.

For instance, the LCM of 4 and 6 is 12, because 12 is the first number that both 4 and 6
can divide. You will find the LCM of 2 numbers. You will set the variables, then initialize
a while loop with an iterator and a Boolean that is True by default. You will set up a
conditional that will break if the iterator divides both numbers. You will increase the
iterator and print the results after the loop completes.

In this activity, using the following steps, you need to find the LCM of 24 and 36.
The steps are as follows:
1. Set a pair of variables as equal to 24 and 36.
2. Initialize the while loop, based on a Boolean that is True by default, with an
iterator.
3. Set up a conditional to check whether the iterator divides both numbers.
4. Break the while loop when the LCM is found.
5. Increment the iterator at the end of the loop.
6. Print the results.

In [13]:
counting = True
x = 24
y = 36
z = 37
while counting:
  if(z%x == 0 and z%y==0):
    break
  z += 1

print("The LCM is ", z)

The LCM is  72


The result is 72

Numbers $x=24$, $y=36$. I have to search $z=37, 38, \ldots$, so that $LCM$ is the smallest value of $z$ such as
$$
z\%x = 0 \\
z\%y = 0
$$

In [14]:
x=24
y=36
z=37

while z>=37 and z<=100:
  if(z%x==0 and z%y==0):
    break
  z+=1

print("The LCM is", z)

The LCM is 72


### For

The for loops are similar to while loops, but they have additional advantages, such as
being able to iterate over strings and other objects.

The for keyword often goes with the in keyword. The i variable is generic. The
phrase, for i in, means that Python is going to check what comes next and look
at its individual components. Strings are composed of characters, so Python
will do something with each of the individual characters. In this particular case,
Python will print out the individual characters, as per the print(i) command.


In [15]:
NamesPython = ["Michelangelo", "Roxane", "Alienor", "Lou", "Arrigo", "Anabelle", "Irene", "Celia", "Carmen", "Oriane", "Sofia"]
print(NamesPython)
for x in NamesPython:
  print(x)

['Michelangelo', 'Roxane', 'Alienor', 'Lou', 'Arrigo', 'Anabelle', 'Irene', 'Celia', 'Carmen', 'Oriane', 'Sofia']
Michelangelo
Roxane
Alienor
Lou
Arrigo
Anabelle
Irene
Celia
Carmen
Oriane
Sofia


In [16]:
N = len(NamesPython)
for i in range(0,N):
  print(NamesPython[i])


Michelangelo
Roxane
Alienor
Lou
Arrigo
Anabelle
Irene
Celia
Carmen
Oriane
Sofia


What if we want to do something with a range of numbers? Can for loops be used
for that? Absolutely. Python provides another keyword, range, to access a range
of numbers. range is often defined by two numbers, the first number, and the last
number, and it includes all numbers in between. Interestingly, the output of range
includes the first number, but not the last number. You will see why in a minute

In [19]:
for x in my_name:
  print(x)

M
i
c
h
e
l
a
n
g
e
l
o


In [18]:
my_name="Michelangelo"
N = len( my_name )
for i in range(0,12):
  print(my_name[i])

M
i
c
h
e
l
a
n
g
e
l
o


Now let's say that you want to count by increments of 2. You can add a third
bound, a step increment, to count up or down by any number desired.
Use a step increment to count the even numbers through 10:

In [20]:
for i in range(1, 11, 2):
  print(i)

1
3
5
7
9


**Example: Demean**

In [26]:
ages = [21, 18, 19, 14, 23, 31, 12, 64, 21, 19, 25]

In [23]:
mean = sum(ages) / len(ages) 
print(mean)



24.272727272727273


In [ ]:
newages = ages
newages[0] = 442
print(newages)
print(ages)

In [ ]:
ages = [21, 18, 19, 14, 23, 31, 12, 64, 21, 19, 25]

In [27]:
newages = ages.copy()
newages[0] = 442
print(newages)
print(ages)


[442, 18, 19, 14, 23, 31, 12, 64, 21, 19, 25]
[21, 18, 19, 14, 23, 31, 12, 64, 21, 19, 25]


In [28]:
newages = ages[:]
newages[0] = 442
print(newages)
print(ages)

[442, 18, 19, 14, 23, 31, 12, 64, 21, 19, 25]
[21, 18, 19, 14, 23, 31, 12, 64, 21, 19, 25]


In [29]:
for i in range(0,len(ages)):
  newages[i] = ages[i] - mean

print(newages)

[-3.2727272727272734, -6.272727272727273, -5.272727272727273, -10.272727272727273, -1.2727272727272734, 6.727272727272727, -12.272727272727273, 39.72727272727273, -3.2727272727272734, -5.272727272727273, 0.7272727272727266]


In [ ]:
mean = sum(ages) / len(ages) 
print(mean)
ages_demean = [(x - mean) for x in ages]

print(ages_demean)

24.272727272727273
[-3.2727272727272734, -6.272727272727273, -5.272727272727273, -10.272727272727273, -1.2727272727272734, 6.727272727272727, -12.272727272727273, 39.72727272727273, -3.2727272727272734, -5.272727272727273, 0.7272727272727266]


## Functions

In the context of programming, a function is a named sequence of statements that
performs a computation. When you define a function, you specify the name and
the sequence of statements. Later, you can “call” the function by name. We have
already seen one example of a function call:





In [ ]:
print("Hello World!")
type(2)

The name of the function is **type** (**print**). The expression in parentheses is called the
**argument** of the function. The argument is a value or variable that we are passing
into the function as input to the function. The result, for the type function, is the
type of the argument.
It is common to say that a function “takes” an argument and “returns” a result.
The result is called the **return value**.

Here are a few other advantages of using Python functions:


*   Avoid repetition of codes
*   Increases program readability
*   Divide a complex problem into simpler ones
*   Reduces chances of error
*   Modifying a program becomes easier by using a function


### Types of functions

There are two types of functions in Python.

 

1. **Built-in Functions** or pre-defined functions
These are the functions which are already defined by Python. For example: id () , type(), print (), casting functions, etc.

 

2. **User-Defined Functions** or Customized Functions
These are functions that are defined by the users for simplicity and to avoid repetition of code.


## Defining a Function

Here’s a simple function named greet_user() that prints a greeting:

In [30]:
def greet_user():
 """Display a simple greeting."""
 print("Hello!")


This example shows the simplest structure of a function. The first line
uses the keyword `def` to inform Python that you’re defining a function. This
is the function definition, which tells Python the name of the function and, if
applicable, what kind of information the function needs to do its job. The
parentheses hold that information. In this case, the name of the function
is `greet_user()`, and it needs no information to do its job, so its parentheses
are empty. (Even so, the parentheses are required.) Finally, the definition
ends in a colon.

Any indented lines that follow def `greet_user()`: make up the body of
the function. The text at line 2 is a comment called a **docstring**, which describes
what the function does. Docstrings are enclosed in triple quotes, which
Python looks for when it generates documentation for the functions in your
programs.

The line print("Hello!") is the only line of actual code in the body
of this function, so `greet_user() `has just one job: print("Hello!").

When you want to use this function, you call it. A function call tells
Python to execute the code in the function. To call a function, you write
the name of the function, followed by any necessary information in parentheses,
as shown below. Because no information is needed here, calling our
function is as simple as entering `greet_user()`. As expected, it prints Hello!:

In [31]:
greet_user()

Hello!


In [32]:
help(greet_user)

Help on function greet_user in module __main__:

greet_user()
    Display a simple greeting.



## Arguments of a function

in the previous example we did not need tl pass any information to the function: the function had no arguments. However, we might need to print the name of the user:

In [ ]:
def greet_user_name(Prenom):
  print("Hello, " + Prenom + "!")

greet_user_name("Michelangelo")

Hello, Michelangelo!


In this case we have one argument, so we have to pass the information to the function.

How about including your full name using 2 arguments?


In [ ]:
def greet_user_full_name(Prenom, Nom):
  print("Hello, " + Prenom + " " + Nom + "!")

In [ ]:
greet_user_full_name("Michelangelo", "Arranz")

Hello, Michelangelo Arranz!


The function understands the arguments given the order (**positional arguments**). But we can pass the arguments explicitly (**keyword arguments**).

During function call, values passed through arguments need not be in the order of parameters in the function definition. This can be achieved by keyword arguments. But all the keyword arguments should match the parameters in the function definition.

In [ ]:
greet_user_full_name(Nom = "Arranz", Prenom = "Michelangelo")

Hello, Michelangelo Arranz!


Sometimes we want to have some arguments with default values: **default arguments**

In [33]:
def greet_user_full_name_plus(Prenom, Nom, excl = True):
  if excl:
    print("Hello, " + Prenom + " " + Nom + "!")
  else:
    print("Hello, " + Prenom + " " + Nom)

In [ ]:
greet_user_full_name_plus("michelangelo", "Arranz")
greet_user_full_name_plus("michelangelo", "Arranz", excl = True)
greet_user_full_name_plus("michelangelo", "Arranz", excl = False)

Hello, michelangelo Arranz!
Hello, michelangelo Arranz!
Hello, michelangelo Arranz




*   default arguments are values that are provided while defining functions.

*   The assignment operator = is used to assign a default value to the argument.
*   default arguments become optional during the function calls
*   If we provide value to the default arguments during function calls, it overrides the default value.
*   The function can have any number of default arguments
*   Default arguments should follow non-default arguments.





**Remember**



1.   default arguments should follow non-default arguments
2.   keyword arguments should follow positional arguments
3.   All the keyword arguments passed must match one of the arguments accepted by the function and their order is not important.
4.   No argument should receive a value more than once
5.   Default arguments are optional arguments



### Sample Mean

In [34]:
ages = [21, 18, 19, 14, 23, 31, 12, 64, 21, 19, 25]
S = sum(ages)
print("The summatory is ", S)
N = len(ages)
m = S/N
print("The average is ", m)

The summatory is  267
The average is  24.272727272727273


In [35]:
def my_mean(x):
  return sum(x)/len(x)

In [43]:
print(my_mean(ages))
m = my_mean(ages)
print(f"La media es {m}")

24.272727272727273
La media es 24.272727272727273


### Exercise: Standard Deviation

The variance is calculated as:

$$
\sigma^2 = \frac{(\sum(X_i - \bar{X})^2}{N-1}
$$

The standard deviation is
$$
\sigma = \sqrt{\sigma^2}
$$

You may call the function `my_mean` inside the `my_sigma` function.

In [45]:
# Calculate the standard deviation of a list of numbers

def my_stdev(x):
  from math import sqrt
  m = my_mean(x)
  xm2 = [(i-m)**2 for i in x]
  s2 = sum(xm2)/len(x)
  return sqrt(s2)

In [46]:
print(my_stdev(ages))

13.477866765997632


In [47]:
import numpy as np
print(np.std(ages))

13.477866765997632


### Arbitrary Arguments

For arbitrary positional argument, an **asterisk (*)** is placed before a parameter in function definition which can hold non-keyword variable-length arguments. These arguments will be wrapped up in a **tuple**. Before the variable number of arguments, zero or more normal arguments may occur.

In [2]:
def add(*x):
  result = 0
  for i in x:
    result += i
  return result

In [3]:
print (add(1,2,3,4,5))

15


## Returns of a function

The `return` statement terminates the function call and returns a value to the caller.

The `return` statement without an expression argument returns `None`, and a function without a return statement also returns `None`.

If the return statement is not reached in the function, it returns `None`.

In [48]:
def add(a,b):
    c=a+b
    if c>10:
        return c
print (add(1,2))

None


**Returning multiple values using commas**

If the expression list in the return statement contains at least one comma, except when part of a list or set display, it returns a tuple. The length of the tuple is the number of expressions in the list. The expressions are evaluated from left to right.

The return type is a tuple.

In [49]:
def add(a,b):
   result=a+b
   return a,b,result
print (add(1,2))

(1, 2, 3)


The return statement having a single expression with a trailing comma returns a tuple.

In [50]:
def add(a,b):
   result=a+b
   return result,
print (add(1,2))

(3,)


**Returning multiple values using list displays**

A list display yields a new list object, the contents being specified by a list of expressions. When a comma-separated list of expressions is supplied, its elements are evaluated from left to right and placed into the list object in that order. The return type is a list.

In [51]:
def calc(a,b):
    sum=a+b
    dif=a-b
    mul=a*b
    div=a/b
    return [sum,dif,mul,div]

print (calc(5,4))

[9, 1, 20, 1.25]


**Returning a single value from a function**

The return statement having a single expression without a trailing comma doesn’t create a tuple; it returns the value of that expression.


In [ ]:
def add(a,b):
   result=a+b
   return result

print (add(1,2))


print (add("Hello ","Python"))


print (add([1,2],[3,4]))


print (add((1,2),(3,4)))

**Multiple ‘return’ statements in a single function**

The function call is terminated when one of the return statements is reached.

In [52]:
def num(a):
    if a%2==0:
        return "Even Number"
    else:
        return "Odd Number"

In [53]:
print(num(5))
print(num(6))

Odd Number
Even Number
